File: src/main/java/kr/dogfoot/hwpxlib/object/content/section_xml/ParaListCore.java
package kr.dogfoot.hwpxlib.object.content.section_xml;

import kr.dogfoot.hwpxlib.object.common.SwitchableObject;
import kr.dogfoot.hwpxlib.object.content.section_xml.paragraph.Para;

import java.util.ArrayList;

public abstract class ParaListCore extends SwitchableObject {
    private final ArrayList<Para> paraList;

    public ParaListCore() {
        paraList = new ArrayList<Para>();
    }

    public int countOfPara() {
        return paraList.size();
    }

    public Para getPara(int index) {
        return paraList.get(index);
    }

    public int getParaIndex(Para para) {
        int count = paraList.size();
        for (int index = 0; index < count; index++) {
            if (paraList.get(index) == para) {
                return index;
            }
        }
        return -1;
    }

    public void addPara(Para para) {
        paraList.add(para);
    }

    public Para addNewPara() {
        Para para = new Para();
        paraList.add(para);
        return para;
    }

    public void insertPara(Para para, int position) {
        paraList.add(position, para);
    }

    public void removePara(int position) {
        paraList.remove(position);
    }

    public void removePara(Para para) {
        paraList.remove(para);
    }

    public void removeAllParas() {
        paraList.clear();
    }

    public Iterable<Para> paras() {
        return paraList;
    }

    protected void copyFrom(ParaListCore from) {
        for (Para para : from.paraList) {
            paraList.add(para.clone());
        }

        super.copyFrom(from);
    }
}

File: js/src/object/content/section_xml/ParaListCore.js
import { SwitchableObject } from '../../common/SwitchableObject.js';
import { Para } from './paragraph/Para.js';

/**
 * Base class for objects that contain a list of paragraphs
 */
export class ParaListCore extends SwitchableObject {
  constructor() {
    super();
    this._paraList = [];
  }

  countOfPara() {
    return this._paraList.length;
  }

  getPara(index) {
    return this._paraList[index];
  }

  getParaIndex(para) {
    return this._paraList.indexOf(para);
  }

  addPara(para) {
    this._paraList.push(para);
  }

  addNewPara() {
    const para = new Para();
    this._paraList.push(para);
    return para;
  }

  insertPara(para, position) {
    this._paraList.splice(position, 0, para);
  }

  removePara(positionOrPara) {
    if (typeof positionOrPara === 'number') {
      this._paraList.splice(positionOrPara, 1);
    } else {
      const index = this._paraList.indexOf(positionOrPara);
      if (index !== -1) {
        this._paraList.splice(index, 1);
      }
    }
  }

  removeAllParas() {
    this._paraList = [];
  }

  paras() {
    return this._paraList;
  }

  /**
   * Gets all text from all paragraphs
   * @param {string} separator - Separator between paragraphs (default: newline)
   * @returns {string}
   */
  getText(separator = '\n') {
    return this._paraList
      .map(para => para.getText())
      .join(separator);
  }

  _copyFromParaListCore(from) {
    this._paraList = [];
    for (const para of from._paraList) {
      this._paraList.push(para.clone());
    }
    this._copyFromSwitchable(from);
  }
}

export default ParaListCore;
